src/services/data-targets/fiware-data-target.service.ts

```diff
@@ -9,9 +10,53 @@ import { DataTargetSendStatus } from "@interfaces/data-target-send-status.interf
 import { FiwareDataTargetConfiguration } from "@interfaces/fiware-data-target-configuration.interface";
 import { BaseDataTargetService } from "@services/data-targets/base-data-target.service";
 
+type TokenEndpointResponse = {
+    access_token: string,
+    expires_in: number,
+}
+
+@Injectable()
+export class AuthenticationTokenProvider {
+
+    private readonly logger = new Logger(AuthenticationTokenProvider.name);
+
+    constructor(private httpService: HttpService, @Inject(CACHE_MANAGER) private cacheManager: Cache) {
+    }
+
+    async getToken(config: FiwareDataTargetConfiguration): Promise<string> {
+        const key = config.clientId + config.clientSecret;
+        const token = await this.cacheManager.get<string>(key)
+        if (token) {
+            this.logger.debug('Token found')
+            return token
+        } else {
+            try {
+                const buffer = Buffer.from(`${config.clientId}:${config.clientSecret}`);
+                const encodedCredentials = buffer.toString('base64');
+                const params = new URLSearchParams([['grant_type', 'client_credentials']])
+                const { data }: { data: TokenEndpointResponse } = await this.httpService.post(config.tokenEndpoint, params, {
+                    headers: {
+                        'Authorization': `Basic ${encodedCredentials}`,
+                        'Content-Type': 'application/x-www-form-urlencoded',
+                    },
+                }).toPromise()
+
+                const clockSkew = 30
```

Comment:
Any reason for adding clock skew? And why exactly is it 30?

Reply:
The reason for adding clock skew is that there might be time small time difference between the clock on the identity server that returns token and server that hosts backend. Also there will be some time lost on communication. In terms of user session usually clock skew is set to 5 minutes but in our scenario token lifetime was only 5 minute so in this case we decided that 30 seconds will be sufficient.

Reply:
I agree with the explanation of clock skew. Here, however, `clockSkew` is only used as an offset for how long before the token expires in the cache. To avoid confusion, rename it to `ttlOffset` or similar

Reply:
Don't you think that `clockSkew` is better name as it explains reasons for that, while `ttlOffset` doesn't explain why you want this offset?

Reply:
I think it's okay. It's only used in this specific context. To avoid confusion in the future, can you add a comment to it with your explanation?
Just to clarify: From my view, clock skew is used to specify by how much a token's timestamps can drift before it's considered invalid. I wouldn't say that it's the cache's responsibility.

Reply:
Done